Select symbol overload in tree unpickling

lrytz opened the pull request:
References to external symbols are pickled as owner + name, so overloads cannot be resolved. This only affects tree unpickling, i.e., only annotation arguments.

At some point in history, the unpickler was using
`Infer.inferMethodAlternative`, but not anymore since fff93cd049.

This commit performs overloading resolution based on the Select tree's type instead: when unpickling `t.f`, the `tpe` of the tree is unpickled. This type can be used to select the correct overload.

Fixes https://github.com/scala/bug/issues/13087

lrytz committed:
Select symbol overload in tree unpickling

References to external symbols are pickled as owner + name, so
overloads cannot be resolved. This only affects tree unpickling,
i.e., only annotation arguments.

At some point in history, the unpickler was using
`Infer.inferMethodAlternative`, but not anymore since fff93cd049.

This commit performs overloading resolution based on the Select
tree's type instead: when unpickling `t.f`, the `tpe` of the tree
is unpickled. This type can be used to select the correct overload.

lrytz commented on src/reflect/scala/reflect/internal/pickling/UnPickler.scala:
```diff
@@ -526,15 +526,6 @@ abstract class UnPickler {
-      def fixApply(tree: Apply, tpe: Type): Apply = {
-        val Apply(fun, args) = tree
-        if (fun.symbol.isOverloaded) {
-          fun setType fun.symbol.info
-          inferMethodAlternative(fun, args map (_.tpe), tpe)
```
This was actually making things worse. Tree types are pickled, so `fun.tpe` is a non-overloaded type before this code. `inferMethodAlternative` is a no-op.

So in unpickled ASTs, a Select of an overloaded method had an overloaded `selectTree.tpe`. Trying to re-pickle such a tree fails (https://github.com/scala/bug/issues/13087).

Removing this coulde would be enough to fix the bug, but I added the new `fixOverload` anyway.

som-snytt commented:
The "balloon payment" commit is very frustrating when examining history. I think a different economic metaphor is needed, perhaps, "permanent lien".

lrytz committed:
Don't crash pickling Literal trees with a null `tpe`

Constant annotation arguments are `Literal(Constant(v))`
ASTs with a non-null `tpe` assigned by typer.

But it's relatively easy to create an untyped Literal
in a compiler plugin, the factory `Literal(Constant(v))`
returns a tree with a `null` type.

This commit prevents an NPE in pickling for this case.

SethTisue commented:
("internal" because the regression never shipped)
